### ISSUE TITLE
IMAP: metadata garbled when envelope spread over 2 lines

### ISSUE THREAD

**Issue by KevinCarhart** [curl/curl#7284]
Hi Daniel et. al.,
I'm referencing #5965 from last September.  The question at the time was, does #5961 fix this?  We have evidence that it doesn't, and we can still raise the problem with libcurl versions past 7.73.0.  I have confirmed it on a newly instantiated cloud environment, so it has very few potential confounding causes.

Karl sends this:
Sometimes an email has an imap envelope across two lines, rather than one.
First line ends in {nnn}, and the first nnn bytes of the second line is the subject.
curl imap only presents the first line through its CURLOPT_HEADERFUNCTION.
(Fortunately, both lines, and the surrounding protocol, can be seen through CURLOPT_DEBUGFUNCTION.)
Our application can't present a proper envelope with just the first line. Subject and other information are not seen.
We need both lines reported through CURLOPT_HEADERFUNCTION.

I haven't found a way to reliably produce a 2 line envelope at the server, but they happen from time to time.

This was originally #5965, suggested that #5961 might fix it, but it does not.

Karl Dahlke

-------------------------------------------------------
Here is my report on what I did on a cloud box.
From your changelog, I saw that #5961 was fixed in 7.73.0, right?

So I wanted to see if it still happened on a curl from more recent than October.  I spun up a pristine ubuntu 21.04 from linode.  Then I issued:

`apt-get install edbrowse`

Verified the libcurl:

`apt-cache policy libcurl4`

`Installed: 7.74.0-1ubuntu2`

Then I configured edbrowse with the server strings and other strings for imap, and confirmed this to be working fine.

When I log in, go to my inbox, and read the froms/subjects I sporadically see messages with _0_ instead of a subject.

Thanks
Kevin

_Originally posted by @KevinCarhart in https://github.com/curl/curl/issues/5965#issuecomment-864530242_

**Comment by bagder** [comment]
Can you perhaps provide a raw dump of the exact contents the server sends that confuses curl? I would like to reproduce the case using the curl test suite!

**Comment by KevinCarhart** [comment]
Yes, here is a failure surrounded by two successes for context.

```
curl>
A018 FETCH 2561 ALL
curl<
* 2561 FETCH (FLAGS (\Seen $X-ME-Annot-2) INTERNALDATE "20-Jun-2021 01:15:54 -0400" RFC822.SIZE 10637 ENVELOPE ("Sat, 19 Jun 2021 22:15:19 -0700" "the message newer than this point will produce a zero" (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) NIL NIL NIL "<35943a04-ad36-48fc-8773-f547893d8933@www.fastmail.com>"))
A018 OK Completed (0.000 sec)
curl>
A019 FETCH 2562 ALL
curl<
* 2562 FETCH (FLAGS (\Seen $X-ME-Annot-2) INTERNALDATE "20-Jun-2021 01:16:19 -0400" RFC822.SIZE 10651 ENVELOPE ("Sat, 19 Jun 2021 22:15:46 -0700" {66}
Save 90% off baby, kids, clothes, toys, accessories and much more! (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) NIL NIL NIL "<76b16ca2-51e7-4f60-a911-f0ef71d94330@www.fastmail.com>"))
A019 OK Completed (0.000 sec)
curl>
A020 FETCH 2563 ALL
curl<
* 2563 FETCH (FLAGS (\Seen $X-ME-Annot-2) INTERNALDATE "20-Jun-2021 02:47:06 -0400" RFC822.SIZE 9378 ENVELOPE ("Sun, 20 Jun 2021 02:46:58 -0400" "encoded subject" (("Karl Dahlke" NIL "[redacted]" "[redacted]")) (("Karl Dahlke" NIL "[redacted]" "[redacted]")) (("Karl Dahlke" NIL "[redacted]" "[redacted]")) ((NIL NIL "kevin_carhart" "fastmail.fm")) NIL NIL NIL "<[redacted]>"))
A020 OK Completed (0.000 sec)
```

**Comment by bagder** [comment]
I presume you mean this block is the problematic one?
~~~
* 2562 FETCH (FLAGS (\Seen $X-ME-Annot-2) INTERNALDATE "20-Jun-2021 01:16:19 -0400" RFC822.SIZE 10651 ENVELOPE ("Sat, 19 Jun 2021 22:15:46 -0700" {66}
Save 90% off baby, kids, clothes, toys, accessories and much more! (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) (("Kevin Carhart" NIL "kevin_carhart" "fastmail.fm")) NIL NIL NIL "<76b16ca2-51e7-4f60-a911-f0ef71d94330@www.fastmail.com>"))
A019 OK Completed (0.000 sec)
~~~

How is this an "envelope spread over 2 lines" ? I see one untagged response line, the body and then finally the end of the server response in the A019 line.

**Comment by KevinCarhart** [comment]
The body is actually blank in this message.  Those 66 characters beginning with 'Save' are the subject line.

**Comment by bagder** [comment]
I meant protocol body, the payload. Not meta data.

To me this looks like a broken IMAP response. since the `* 2562` line indicates that there are 66 bytes coming and then there's instead 354 bytes sent before the final `A019 OK` line.

The title of this issue says that the "envelope" is "spread over 2 lines" but A) I don't see that in the protocol dump and B) I don't see in the IMAP spec how that is supported by the protocol. Can you help me understand that?

**Comment by KevinCarhart** [comment]
Handing over to Karl:
I'm not entirely sure of the technical definition of an envelope,
but I do know that when it isn't wholly contained in the first line, we need the second line to complete the envelope.
A continuation of the envelope information is indicated, on the first line, by a number in braces.
That is the length of the subject, which is in this case not quoted, because we already know how long it is.
It is on the second line, along with recipients etc.
Our application, or any application, would need the second line to present the envelope to the user,
whereupon he can decide what to do with the email.
So far we have not asked for the body of the email, just the envelope.
The first line is delivered to us by CURLOPT_WRITEFUNCTION callback method.
If that first line is the entire envelope we are fine.
And it usually is.
But if not, we never see the second line, never see the subject or recipients or reference id etc.
We can only present a blank line to the user, whence he has to fetch the entire email to see what it is.

There is a workaround, which we perform, wherein we use the CURLOPT_DEBUGFUNCTION to monitor the entire conversation, and thus grab the second line, and paste it onto the first, to produce a proper envelope.
This is rather awkward, and not easy to maintain.

I believe that, whatever you might call the second line, a matter of semantics, it should be provided to the application through CURLOPT_WRITEFUNCTION, just as the first.
Unless there is another way to see this second line, without going through CURLOPT_DEBUGFUNCTION methods, that I am not aware of.

Karl Dahlke

**Comment by bagder** [comment]
Sorry, but I still don't understand how the shown protocol snippet is supposed to work.

As I understand the IMAP protocol, response data from the server can only be tagged (it starts with the `A00*` thing), untagged (it starts with `* `) or "other" (everything else).

The size of "everything else" is the number of bytes within the braces on the previous untagged response line. In this particular case **66** bytes.

Are you saying that what comes after those 66 bytes is a continuation of the untagged response line?

**Comment by KevinCarhart** [comment]
From Karl:
Yes.
The point is, we need the "other".
The first 66 bytes is the subject, which is part of the envelope, then the rest of the line is the rest of the envelope, recipients, cc, reference id, etc.
Usually the untagged line holds the entire envelope, but not always, and when it doesn't, we need the other to put it all together.

**Comment by bagder** [comment]
Aaaah yes. Now it all comes back to me. Thanks!

**Comment by bagder** [comment]
I created test case 897 in #7748 where I added more data to that header and in the test case verify that curl delivers it.

IOW: I could not reproduce this bug.

**Comment by KevinCarhart** [comment]
Thank you, this seemed to enable us to push a change that fixes it.  We appreciate the help.
Karl writes:
A year ago, we implemented an imap client in edbrowse, using the curl library.
(See edbrowse.org for more information.)
After FETCH nnn ALL, we used WRITEFUNCTION to grab the untagged line, which holds the envelope.
We then presented the envelope information to the user,
whence he could read the entire email, delete it, move it to another folder, skip it, save it to a file, etc, all through the command line.
This worked great, and I didn't think much more about it,
but then servers started generating envelopes across 2 lines, the untagged line and a followup line.
Not often, maybe a few percent, but it was annoying because we couldn't get and parse the envelope data,
and the user had to read the actual email just to know who it was from, date, size, etc.
I thought perhaps WRITEFUNCTION was not giving me the second line, and so I started this chain of inquiry.
I appreciate your patience, and your work on our behalf.

As you explained, HEADERFUNCTION must be used to grab the second line, if there is one.
I now use WRITEFUNCTION and HEADERFUNCTION in proper combination,
and I get the entire envelope, whether it is on one line or two.

Again, thank you for helping us, and for writing a test case, to ensure this continues to work properly in the future.

I can't imagine edbrowse without the curl library, to implement the internet transport layer.
We couldn't possibly write all that software ourselves.

Sincerely,

Karl Dahlke
